### PR TITLE
Add GKE recipe

### PR DESCRIPTION
**Pull request by camphillips22**
This recipe provides a script that deploys a recommended
hauser deployment to GKE, including a service account
with "least privilege".

### CODE REVIEW

File: recipes/GKE Setup/README.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+```
+
+To verify that everything is working as expected, you can also check the logs with
+```shell
+kubect logs -lapp=hauser -c hauser
+```
+
+## Service accounts
+This setup is based on using a service account for the hauser deployment.
+In short, this is recommended to adhere to the principal of east privilege.
```
Comment (jhump) [inline recipes/GKE Setup/README.md:88]:
```suggestion
In short, this is recommended to adhere to the principal of least privilege.
```

File: recipes/GKE Setup/README.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+```shell
+BIGQUERY_DATASET="fullstory-export-data"
+SERVICE_NAME="hauser-svc"
+GCS_BUCKET="$GOOGLE_PROJECT-hauser"
+HAUSER_VERSION="latest"
+```
+To find your FullStory Org Id, see [help docs](https://help.fullstory.com/hc/en-us/articles/360047075853-How-do-I-find-my-FullStory-Org-Id-).
+You will also need to create or get a [FullStory API Key](https://help.fullstory.com/hc/en-us/articles/360020624834-Where-can-I-find-my-API-key-).
+
+The `HAUSER_VERSION` variable corresponds to the docker image tag. See [docker hub](https://hub.docker.com/r/fullstorydev/hauser) for available tags.
```
Comment (jhump) [inline recipes/GKE Setup/README.md:53]: Maybe mention that the latest available version will be used by default?

Comment (jhump) [reply recipes/GKE Setup/README.md:53]: Also, might move this up above to line 40-ish. That way you introduce all of the parameters and their defaults first, and then the following paragraph describes how to change them by editing values in the script.

File: recipes/GKE Setup/README.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+## The deploy script
+The `deploy.sh` script in this directory uses the gcloud SDK and associated CLI's to
+create a service account, add the permissions, and create a Kubernetes deployment that
+runs hauser.
+
+Assuming you have completed the prerequisites, you can run the deploy script with:
+```shell
+./deploy.sh -p "<my_project_id>" -o "<my_org_id>" -k "<fs_api_key"
```
Comment (jhump) [inline recipes/GKE Setup/README.md:62]: It's usually bad practice to stick secrets in command-line args. Perhaps this should accept the API key via env var instead, and then complain at script startup if the env var is unset?

(Also, nit, missing a trailing `>`.)

File: recipes/GKE Setup/README.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+## Service accounts
+This setup is based on using a service account for the hauser deployment.
+In short, this is recommended to adhere to the principal of east privilege.
+To read more about service accounts in Google, see [Google's documentation](https://cloud.google.com/iam/docs/service-accounts)
+
+The `create-svc-account.sh` script will create/validate that the service account
+exists and has all of the necessary permissions and IAM policies associated with it.
+This includes:
+* [Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) to associate the gcloud
```
Comment (jhump) [inline recipes/GKE Setup/README.md:94]: Do we need content about how to configure the workload if they are _not_ using workload identity?

That is a bit more involved, unfortunately, requiring one to provision a JSON key, upload it as a k8s secret and then edit the deployment to mount the secret and then point to the file therein via `GOOGLE_APPLICATION_CREDENTIALS` env var.

File: recipes/GKE Setup/create-svc-account.sh [hunk 5]
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+
+set -e
+
+SERVICE_EMAIL="$SERVICE_NAME@$GOOGLE_PROJECT.iam.gserviceaccount.com"
```
Comment (jhump) [inline recipes/GKE Setup/create-svc-account.sh:5]: The README should probably mention that `SERVICE_NAME` must be set before invoking. Or perhaps this should accept it as an argument?

Also, this script should definitely validate that it's not empty before proceeding.

Comment (jhump) [reply recipes/GKE Setup/create-svc-account.sh:5]: `GOOGLE_PROJECT`, `BIGQUERY_DATASET`, and `GCS_BUCKET`, too

File: recipes/GKE Setup/README.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+```shell
+kubect logs -lapp=hauser -c hauser
+```
+
+## Service accounts
+This setup is based on using a service account for the hauser deployment.
+In short, this is recommended to adhere to the principal of east privilege.
+To read more about service accounts in Google, see [Google's documentation](https://cloud.google.com/iam/docs/service-accounts)
+
+The `create-svc-account.sh` script will create/validate that the service account
```
Comment (jhump) [inline recipes/GKE Setup/README.md:91]: It doesn't look like users are intended to run this directly, but instead just let `deploy.sh` run it. Is that right? Maybe merge into `deploy.sh` if that is the case? Or document it a little more (required inputs/params) and/or make it more usable to be run independently?

File: recipes/GKE Setup/deploy.sh [hunk 7]
```diff
@@ -0,0 +1,104 @@
+#!/bin/bash
+
+set -e
+
+GOOGLE_PROJECT=""
+ORG_ID=""
+BIGQUERY_DATASET="fullstory-export-data"
+SERVICE_NAME="hauser-svc"
+GCS_BUCKET="$GOOGLE_PROJECT-hauser"
```
Comment (jhump) [inline recipes/GKE Setup/deploy.sh:9]: I think the script needs to be tested in an empty/dummy project. For example,`GOOGLE_PROJECT` var is empty at this point, which means this script cannot work.

Comment (jhump) [reply recipes/GKE Setup/deploy.sh:9]: You could remove the command-line args and make the interface entirely driven by env vars. Then you can rely on `GOOGLE_PROJECT` being set (since the script will abort if it is not). To preserve defaults for the ones below, you can do this:
```sh
BIGQUERY_DATASET="${BIGQUERY_DATASET:-fullstory-export-data}"
```

File: recipes/GKE Setup/hauser.yaml [hunk 8]
```diff
@@ -0,0 +1,72 @@
+---
```
Comment (jhump) [inline recipes/GKE Setup/hauser.yaml:1]: I wonder if this would be better to provide as a helm chart and vars file and a terraform module...

@nrmitchi and @bschaeffer may have thoughts.

File: recipes/GKE Setup/README.md (hunk 3 above)
Comment (camphillips22) [reply recipes/GKE Setup/README.md:62]: Yeah, I was vacillating on that. It actually already works if you don't provide it. I'll just remove it from the args and require that it's set before running the script.

File: recipes/GKE Setup/README.md (hunk 4 above)
Comment (camphillips22) [reply recipes/GKE Setup/README.md:94]: Yeah, this is probably the part I'm least clear about. Just so I understand, adding the service account annotation will not work if WID is not enabled? I was going to do the whole "create a json key and add it to the secret" thing at first, but went this route first since it's a bit simpler. Maybe we prompt to either use WID or use private key directly?

File: recipes/GKE Setup/README.md (hunk 6 above)
Comment (camphillips22) [reply recipes/GKE Setup/README.md:91]: Yeah, I think I might just merge it. That'll definitely simplify the validation bits.

File: recipes/GKE Setup/hauser.yaml (hunk 8 above)
Comment (nrmitchi) [reply recipes/GKE Setup/hauser.yaml:1]: I would agree. A Helm chart here would typically accept a service account to use as a parameter, and the GCP service account would be linked elsewhere. This would allow the basic kubernetes definitions/deployment to be applicable in any kubernetes cluster, and not specific to GKE.

Comment (camphillips22) [reply recipes/GKE Setup/hauser.yaml:1]: I definitely think there's value in adding these, but I wanted to start with a "minimal set" of deps (just gcloud and jq) so that I could get a working GKE recipe quicker. Also, I'm not as familiar with setting up the helm chart (typically in it's own repo?) and I haven't worked with the other clouds enough to know how to write that more generally.

Comment (camphillips22) [reply recipes/GKE Setup/hauser.yaml:1]: But maybe it's simpler than I'm imagining. Just haven't done the research here yet.

Comment (nrmitchi) [reply recipes/GKE Setup/hauser.yaml:1]: I don't think a Helm chart that encompasses what you have here would be difficult; it would mostly be wrapping the manifests you've already written, and updating the documentation.

I don't think the lack-of-helm should be considered a blocker here, but we should come back and add it. This is because hauser is safe to remove and recreate, which would make a switch to helm easier for anyone using this.
